### PR TITLE
Instrument usage of include_flutter.groovy and xcode_backend.sh

### PR DESCRIPTION
**Pull request by blasten**
## Description

Instrument usage of `include_flutter.groovy` and `xcode_backend.sh` via `flutter build bundle`

## Related Issues
https://github.com/flutter/flutter/issues/32649

## Tests

I added the following tests:

`build_bundle_test.dart`

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter_tools/lib/src/bundle.dart [hunk 1]
```diff
@@ -48,6 +48,46 @@ const String _kVMSnapshotData = 'vm_snapshot_data';
 const String _kIsolateSnapshotData = 'isolate_snapshot_data';
 const String _kIsolateSnapshotInstr = 'isolate_snapshot_instr';
 
+class BundleFactory {
```
Comment (xster) [inline packages/flutter_tools/lib/src/bundle.dart]: Not sure I understood this part. What's the intent here?

File: packages/flutter_tools/bin/xcode_backend.sh [hunk 2]
```diff
@@ -249,7 +249,7 @@ BuildApp() {
   fi
 
   StreamOutput " ├─Assembling Flutter resources..."
-  RunCommand "${FLUTTER_ROOT}/bin/flutter" --suppress-analytics             \
+  RunCommand "${FLUTTER_ROOT}/bin/flutter"                                  \
```
Comment (xster) [inline packages/flutter_tools/bin/xcode_backend.sh]: It would be nice to avoid double logging re-entrant calls back into Flutter such as run->xcode_backend.sh->bundle.

Could we set a flag so that only the outermost invoked Flutter CLI command gets analytics and all nested calls don't?

File: packages/flutter_tools/lib/src/bundle.dart (hunk 1 above)
Comment (blasten) [reply packages/flutter_tools/lib/src/bundle.dart]: So, we can mock `buildBundle`. I added a comment.

File: packages/flutter_tools/bin/xcode_backend.sh (hunk 2 above)
Comment (blasten) [reply packages/flutter_tools/bin/xcode_backend.sh]: Done.

File: packages/flutter_tools/lib/src/ios/mac.dart [hunk 3]
```diff
@@ -279,6 +279,10 @@ Future<XcodeBuildResult> buildXcodeProject({
     buildCommands.add('SCRIPT_OUTPUT_STREAM_FILE=${scriptOutputPipeFile.absolute.path}');
   }
 
+  // Don't log analytics for downstream Flutter commands.
+  // e.g. `flutter build bundle`.
+  buildCommands.add('SUPPRESS_ANALYTICS=true');
```
Comment (xster) [inline packages/flutter_tools/lib/src/ios/mac.dart]: We should do this for Android too (maybe somewhere in gradle.dart or apk.dart?)

**Review by xster — COMMENTED** [review]
LGTM

File: packages/flutter_tools/bin/xcode_backend.sh (hunk 2 above)
Comment (zanderso) [reply packages/flutter_tools/bin/xcode_backend.sh]: I think we need a more nuanced solution, here, though not needed for this CL. For example it would be nice to have timing events from re-entrant calls for build steps etc.

File: packages/flutter_tools/lib/src/bundle.dart [hunk 4]
```diff
@@ -48,6 +48,47 @@ const String _kVMSnapshotData = 'vm_snapshot_data';
...
+    String packagesPath,
+    bool precompiledSnapshot = false,
+    bool reportLicensedPackages = false,
+    bool trackWidgetCreation = false,
+    String compilationTraceFilePath,
+    List<String> extraFrontEndOptions = const <String>[],
+    List<String> extraGenSnapshotOptions = const <String>[],
+    List<String> fileSystemRoots,
+    String fileSystemScheme,
+  }) => build(
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/bundle.dart]: Not sure if this will help make things cleaner, but can you try making `build` an instance method of `BundleFactory`. Then I guess `buildBundle` would be the constructor, and you would use it like `BundleFactory(...)..build();` Might help avoid repeating this long parameter list so many times.

File: packages/flutter_tools/lib/src/usage.dart [hunk 5]
```diff
@@ -66,8 +70,9 @@ class Usage {
     }
     _analytics.analyticsOpt = AnalyticsOpt.optOut;
 
+    final bool suppressEnvFlag = platform.environment['SUPPRESS_ANALYTICS'] == 'true';
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/usage.dart]: Consider `FLUTTER_SUPPRESS_ANALYTICS`

File: packages/flutter_tools/bin/xcode_backend.sh (hunk 2 above)
Comment (xster) [reply packages/flutter_tools/bin/xcode_backend.sh]: if we want screen/event/timing time breakdown, we likely want to consider something more nuanced than GA. Otherwise, non session joined nested acts double logged as siblings may not be as actionable as data.

Comment (zanderso) [reply packages/flutter_tools/bin/xcode_backend.sh]: 1. We can pass whatever information we'd like to pass to the re-entrant calls.
2. Logging behavior can be different in a re-entrant context.
We will probably need to do both of those things whether GA is the backend or something else.

File: packages/flutter_tools/lib/src/bundle.dart (hunk 4 above)
Comment (blasten) [reply packages/flutter_tools/lib/src/bundle.dart]: I renamed  `BundleFactory` as `BundleBuilder`.  I also merged `BundleBuilder().build(..)` and `build(..)`.   I thought about moving the parameters to the constructor, but that would make the initialization in `build_bundle.dart` a little complex. e.g.  can't be done in the constructor.  Also, we use a similar pattern in https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/commands/attach.dart#L276-L286

File: packages/flutter_tools/bin/xcode_backend.sh (hunk 2 above)
Comment (blasten) [reply packages/flutter_tools/bin/xcode_backend.sh]: +1

Comment (xster) [reply packages/flutter_tools/bin/xcode_backend.sh]: What I meant by sibling log entries is that

- Yes, we can tag reentrant flutter build bundle to say that their gradle/shell script parent caller's parent was either a flutter command or not.
- But we won't be able to answer the questions of "for all the calls to flutter run that was 30s+, how much time did flutter build bundle take in median" without a self definable schema that allows sessionization beyond what the ga_sessions_* table lets us do. 
- We can repurpose custom dimensions of each top level command to bundle sub-command timing breakdowns as strings but then we're re-inventing the wheel.

In other words, whether GA is the backend or something implies pre-defined logging schema or not.

**Review by zanderso — APPROVED** [review]

File: packages/flutter_tools/bin/xcode_backend.sh (hunk 2 above)
Comment (zanderso) [reply packages/flutter_tools/bin/xcode_backend.sh]: I'm sure that's all true, but I don't think we should predicate collecting data from re-entrant flutter_tool invocations on switching to a new backend. For example, if ~all bundle subcommand invocations come from re-entrant invocations, then we can't make any queries about that information because we don't have it, let alone complex queries.
